### ISSUE TITLE
Arrays and nullable arrays don't decode into the same types

### ISSUE THREAD

**Issue by kklipsch** [linkedin/goavro#90]
https://github.com/kklipsch/goavro.v2/blob/slice/record_test.go#L472-L519

Coming out of the nullable code you get the type you put in (in our test case []int32) coming out of the non-nullable code you get []interface{}.

This seems bad.

**Comment by karrick** [comment]
Yeah, I agree this is not optimal, but I'm not sure a way around it unless we do code generation. The original API was any time you encoded a chunk of data as an array, you passed it a `[]interface{}`, and when you decoded an array you'd get a `[]interface{}` back. It makes sense. Without code generation, there isn't a way around it that I can think of.

However, many engineers were confused by the fact that when you provided a `[]int64` to a codec for encoding that wanted an array of longs, it would complain that the data type was a mismatch. So there is now a convenience function for accepting a `interface{}` and the encoder will, as a convenience, convert to `[]interface{}`, which is what the encoder really wants.

I'm going to mark this Enhancement and Help Wanted for now. Maybe someone else has a better idea.
